File: src/main/java/com/kry/kata/bowling/BowlingGame.java
package com.kry.kata.bowling;

public class BowlingGame {

    private int[] pins = new int[21];
    private int cur = 0;

    public void roll(final int pins) {
        this.pins[cur] = pins;
        this.cur += 1;

    }

    public int score() {
        int score = 0;

        for (int i = 0; i < cur; ) {
            int moveTo = isStrike(i) ? 1 : 2;

            score += calFrameScore(i);

            i += moveTo;
        }

        return score;
    }

    private int calFrameScore(int at) {
        if (isStrike(at)) {
            return 10 + this.pins[at + 1] + this.pins[at + 2];
        }

        final int frameScore = this.pins[at] + this.pins[at + 1];

        if (frameScore == 10) {
            return 10 + this.pins[at + 2];
        }

        return frameScore;
    }

    private boolean isStrike(int i) {
        return this.pins[i] == 10;
    }
}

File: src/main/kotlin/com/kry/kata/bowling/BowlingGame.kt
package com.kry.kata.bowling

class BowlingGame {
    private val pins = Array(21) { 0 }
    private var num = 0

    fun roll(pins: Int) {
        this.pins[num++] = pins
    }

    fun score(): Int {
        var pins = this.pins.toList()
        var boll = 0

        val frames = object : Frames {
            override fun strike(): Boolean = pins[boll] == 10

            override fun appendScoreForStrike(): Int = pins[boll + 1] + pins[boll + 2]

            override fun moveForStrike() {
                boll += 1
            }

            override fun appendScoreForSpare(): Int = pins[boll + 2]

            override fun scoreForFrame(): Int = pins[boll] + pins[boll + 1]

            override fun move() {
                boll += 2
            }
        }

        return cal(frames)
    }

    private fun cal(frames: Frames): Int {
        var score = 0

        repeat(10) r@{
            if (frames.strike()) {
                score += 10 + frames.appendScoreForStrike()
                frames.moveForStrike()
                return@r
            }

            val s = frames.scoreForFrame()
            score += s

            if (s == 10) score += frames.appendScoreForSpare()
            frames.move()
        }

        return score
    }

    interface Frames {
        fun strike(): Boolean

        fun appendScoreForStrike(): Int

        fun moveForStrike()

        fun appendScoreForSpare(): Int

        fun scoreForFrame(): Int

        fun move()
    }
}
